In [1]:
import pandas as pd
import geopandas as gpd
import os
import re
print(os.getcwd())
os.chdir(r"C:\Users\LMENENDEZ\GitHub\MultiModX")
print(os.getcwd())
pd.set_option('display.max_columns', None)

c:\Users\LMENENDEZ\GitHub\MultiModX\notebooks
C:\Users\LMENENDEZ\GitHub\MultiModX


NOTE: I might want to remove the legs column in the future 

Some characteristics of the trips that go abroad:
- if the person is spanish: we know the "roaming out", i.e., the signal they emitted when they connect to roaming. Hence we know with certainty the country where they arrived. However, if they did a connection, we only know the country where they first connected. We have to assume that that is their final destination
- if the person is not spanish, we assume their country of destination is their origin country. 
For trips that come from abroad a similar situation holds

This might create inaccuracies in the results


# 1. Format and eliminate unnecessary trips

In [2]:
# Trips during the week 22/09/2022 28/09/2022 (thursday to thursday)
# the day of study selected was Friday to put the air layer under pressure
all_trips = pd.read_csv(
    r"G:\Unidades compartidas\04_PROYECTOS I+D+i\2023 MultiModX\iii) Project\WP4 Performance Assessment Solution\Demand data\Matrices MITMA\with_archetypes\20220922_28_trip_matrix_arc_pt_processed.csv.gz",
    compression="gzip",
    sep="|"
)

In [3]:
#Here there is no trip id but later trip id appears all the time
trips = all_trips[all_trips["date"] == 20220923].reset_index(drop=True).rename(columns={"origin_nut": "origin", "destination_nut": "destination"})
trips.head()

date trip_period origin_zone origin origin_name destination_zone  \
0  20220923         P00       01002  ES211       Álava          4802006   
1  20220923         P00    01009_AM  ES211       Álava         31010_AM   
2  20220923         P00    01009_AM  ES211       Álava       abroad_208   
3  20220923         P00       01036  ES211       Álava          4802006   
4  20220923         P00     0105902  ES211       Álava            09219   

  destination destination_name entry_point     exit_point origin_purpose  \
0       ES213          Vizcaya         NaN            NaN             NF   
1       ES220          Navarra         NaN            NaN              O   
2      abroad           abroad         NaN  ground_Fra_08             NF   
3       ES213          Vizcaya         NaN            NaN              H   
4       ES412           Burgos         NaN            NaN              O   

  destination_purpose           distance     route_distance duration   mode  \
0                   O  D04_[10000-50000)  D04_[10000-50000)    01-02  train   
1                   H  D04_[10000-50000)  D04_[10000-50000)    00-01  train   
2                  NF             abroad    D05_[50000-inf)    01-02  train   
3                   O  D04_[10000-50000)  D04_[10000-50000)    00-01  train   
4                  NF  D04_[10000-50000)  D04_[10000-50000)    01-02  train   

        service                                               legs  \
0  conv_unknown  P00*01002*01002*None*train_13121*00-01*road*No...   
1  conv_unknown  P00*01009_AM*01009_AM*None*train_11213*00-01*r...   
2  conv_unknown  P00*01009_AM*01009_AM*None*train_11213*00-01*r...   
3  conv_unknown  P00*01036*01036*None*train_13106*00-01*road*No...   
4  conv_unknown  P00*0105902*0105901*None*train_11208*00-01*roa...   

  trip_vehicle_type nationality home_census home_zone overnight_census  \
0             other          ES        2_48   4801303             2_48   
1             other          ES        2_31  31010_AM             2_31   
2             other          FR         NaN       NaN              NaN   
3             other          ES        2_01     01036             2_48   
4             other          ES        2_01   0105904             2_01   

              income          age     sex vehicle_type  \
0  I01_[10000-15000)  A01_[25-45)    male    passenger   
1  I01_[10000-15000)  A01_[25-45)    male    passenger   
2                NaN          NaN     NaN    passenger   
3    I02_[15000-inf)  A02_[45-65)  female    passenger   
4    I02_[15000-inf)  A02_[45-65)    male    passenger   

   short_professional_driver  trips  trips_km  sample_trips  archetype_0  \
0                      False  4.135   139.516           1.0     0.000000   
1                      False  5.191    69.312           1.0     3.707857   
2                      False  1.599   309.588           1.0     0.888333   
3                      False  6.236   139.477           1.0     3.118000   
4                      False  4.215   155.813           1.0     3.417568   

   archetype_1  archetype_2  archetype_3  archetype_4  archetype_5  n_legs  \
0     0.000000     2.067500     2.067500     0.000000     0.000000       3   
1     0.000000     0.000000     0.494381     0.000000     0.988762       3   
2     0.399750     0.142133     0.071067     0.062183     0.035533       3   
3     0.000000     0.779500     1.559000     0.779500     0.000000       3   
4     0.113919     0.227838     0.227838     0.227838     0.000000       3   

     mode_sequence            node_sequence   start_node     end_node  \
0  road-train-road  train_13121-train_13200  train_13121  train_13200   
1  road-train-road  train_11213-train_11300  train_11213  train_11300   
2  road-train-road  train_11213-train_11600  train_11213  train_11600   
3  road-train-road  train_13106-train_13200  train_13106  train_13200   
4  road-train-road  train_11208-train_11200  train_11208  train_11200   

              type  road_legs  train_legs  p

In [4]:
%load_ext autoreload

In [5]:
%autoreload
from script.trips_format import *

In [6]:
#associates each airport to the corresponding new NUTS
airports_to_NUTS={"airport_LPA":("ES705","Gran Canaria"),
                 "airport_FUE":("ES704","Fuerteventura"),
                 "airport_ACE":("ES708","Lanzarote"),
                 "airport_TFS":("ES709","Tenerife"),
                 "airport_TFN":("ES709","Tenerife"),
                 "airport_GMZ":("ES709","Tenerife"),
                 "airport_SPC":("ES707","La Palma"),
                 "airport_VDE":("ES703","El Hierro"),
                 "airport_PMI":("ES532","Mallorca"),
                 "airport_IBZ":("ES531","Eivissa i Formentera"),
                 "airport_MAH":("ES533","Menorca")}

In [7]:
trips=format_trips(trips,airports_to_NUTS)

17 columns were removed


First we remove the cercanías trips and the trips that go through weird stations

In [8]:
# remove cercanías
trips=trips[~(((trips["origin"]=="ES424")&(trips["destination"]=="ES300"))|((trips["origin"]=="ES300")&(trips["destination"]=="ES424")))]

In [9]:
# location of "ALL" train stops given by UiC
# However this list is still incomplete
stops_loc=pd.read_csv(r"G:\Unidades compartidas\04_PROYECTOS I+D+i\2023 MultiModX\iii) Project\WP3 Scenario definition\Case study input data\Spain+abroad\v=0.1\infrastructure\rail_info\stops.txt").astype(str) # everything is a string here to match other formatting
stops_loc["stop_id"] = stops_loc["stop_id"].apply(lambda x: "00" + x) #to make they start with 00

In [10]:
trips.loc[:,"weird_stations"] = trips["node_sequence_reduced"].apply(
    lambda x: find_weird_stations(x, stops_loc))

In [11]:
unique_weird_stations=get_weird_stations(trips["weird_stations"])

In [12]:
len(unique_weird_stations)

414

In [13]:
MobA_stations_coord=gpd.read_file(r"G:\Unidades compartidas\04_PROYECTOS I+D+i\2023 MultiModX\iii) Project\WP3 Scenario definition\Case study input data\Spain\v=0.7\oferta_transporte\train_stations\train_stations.shp")

In [14]:
# identifies all the un-localisable stations
nowhere_stations=set(unique_weird_stations)-set(MobA_stations_coord["ID"])
print(f"there are {len(nowhere_stations)} stations that are not in the data provided by MobA but appear in the trips dataframe")

there are 205 stations that are not in the data provided by MobA but appear in the trips dataframe


In [15]:
trips = trips[~trips["node_sequence_reduced"].apply(lambda x: any(station in x for station in nowhere_stations))]

Then we add the relevant information for international trips

In [16]:
international_codes=pd.read_csv(r"G:\Unidades compartidas\04_PROYECTOS I+D+i\2023 MultiModX\iii) Project\WP3 Scenario definition\Case study input data\Spain+abroad\v=0.1\infrastructure\countries mcc\mcc_to_nationality.txt", sep="|")
international_codes

mcc                        country_name country
0    202                              Greece      GR
1    204                         Netherlands      NL
2    206                             Belgium      BE
3    208                              France      FR
4    212                              Monaco      MC
..   ...                                 ...     ...
227  746                            Suriname      SR
228  748                             Uruguay      UY
229  750         Falkland Islands (Malvinas)      FK
230  901  International & Satellite Networks       -
231  995                      United Kingdom      GB

[232 rows x 3 columns]

In [17]:
mcc_to_country=international_codes.set_index("mcc")["country"].to_dict()

In [18]:
#it is better to divide trips into trips that go to or come from abroad and trips that do not come from
#or go to abroad to save time in computations

trips_abroad=trips[(trips["origin"]=="abroad")|(trips["destination"]=="abroad")]
trips_national=trips[~((trips["origin"]=="abroad")|(trips["destination"]=="abroad"))]

In [19]:
trips_abroad=format_trips_abroad(trips_abroad,mcc_to_country)

In [20]:
trips_national=format_trips_national(trips_national)

In [21]:
trips=pd.concat([trips_national,trips_abroad])

In [22]:
trips

date origin origin_name destination    destination_name  \
0      20220923  ES111    A Coruña       ES112                Lugo   
1      20220923  ES111    A Coruña       ES112                Lugo   
2      20220923  ES111    A Coruña       ES112                Lugo   
3      20220923  ES111    A Coruña       ES112                Lugo   
4      20220923  ES111    A Coruña       ES112                Lugo   
...         ...    ...         ...         ...                 ...   
77395  20220923     ZA          ZA       ES300              Madrid   
77396  20220923     ZA          ZA       ES511           Barcelona   
77397  20220923     ZA          ZA       ES511           Barcelona   
77398  20220923     ZA          ZA       ES521  Alicante / Alacant   
77399  20220923     ZA          ZA       ES617              Málaga   

         entry_point exit_point origin_purpose destination_purpose  \
0                NaN        NaN              H                  NF   
1                NaN        NaN              H                  NF   
2                NaN        NaN              H                  NF   
3                NaN        NaN              H                  NF   
4                NaN        NaN              H                  NF   
...              ...        ...            ...                 ...   
77395    airport_MAD        NaN             NF                  NF   
77396    airport_BCN        NaN             NF                  NF   
77397    airport_BCN        NaN             NF                  NF   
77398  ground_Por_00        NaN             NF                  NF   
77399    airport_AGP        NaN             NF                  NF   

                                                    legs nationality  \
0      P06*15009*1501702*None*train_20402*00-01*road*...          ES   
1      P07*15005*1503009*None*train_20410*00-01*road*...          ES   
2      P07*1503007*1503004*None*train_31412*00-01*roa...          ES   
3      P08*15008*15009*None*train_20400*00-01*road*No...          ES   
4      P08*1503009*1503004*None*train_31412*00-01*roa...          ES   
...                                                  ...         ...   
77395  P19*abroad_655*2807921*None*airport_MAD*00-01*...          AR   
77396  P07*abroad_655*0816904*None*airport_BCN*00-01*...          AR   
77397  P13*abroad_655*0816904*None*airport_BCN*00-01*...          ZA   
77398  P15*abroad_655*3205405*None*train_22100*00-01*...          AR   
77399  P13*abroad_655*2906711*None*airport_AGP*00-01*...          ZA   

       archetype_0  archetype_1  archetype_2  archetype_3  archetype_4  \
0         3.937095     0.051465     0.141530     0.424589     0.257326   
1         3.351054     0.043805     0.120463     0.361388     0.219023   
2         4.821164     0.129254     0.064627     0.348985     0.413612   
3         2.562062     0.033491     0.092100     0.276301     0.167455   
4         7.604907     0.241774     0.109897     0.527508     0.615426   
...            ...          ...          ...          ...          ...   
77395     5.411086     1.365160     1.625308     0.066895     0.304745   
77396     1.372793     1.028286     0.975929     0.048168     0.104713   
77397     1.372793     1.028286     0.975929     0.048168     0.104713   
77398     0.992828     2.109759     0.496414     0.000000     0.000000   
77399     1.462094     1.349625     0.674813     0.000000     0.056234   

       archetype_5  n_legs          mode_sequence  \
0         0.192995       3        road-train-road   
1         0.164267       3        road-train-road   
2         0.284358       3        road-train-road   
3         0.125591       3        road-train-road   
4         0.549487       3        road-train-road   
...            ...     ...                    ...   
77395     0.076806       2             plane-road   
77396     0.069111       4  plane-road-train-road   
77397     0.069111       2             plane-road   
77398     0.000000       3        road-train-

# 2. Analysis of trips going abroad

We can distinguish the region in France and Portugal where the trip finishes, however we do not have the same level of detail with the rest of the countries, so we want to group those trips. 

In [25]:
trips_from_abroad=trips[~(trips["origin"].str.startswith("ES"))]

In [27]:
trips_to_abroad=trips[~(trips["destination"].str.startswith("ES"))]

In [28]:
(trips_abroad["trips"].sum()/trips["trips"].sum())*100

np.float64(33.04858065861245)

33.05% of trips from my matrix either come from abroad or go abroad

In [31]:
trips_to_abroad[trips_to_abroad["destination"]=="DE"]["trips"].sum()

np.float64(4445.545)

In [32]:
outgoing_grouped=trips_to_abroad.groupby("destination")["trips"].sum().reset_index()

In [33]:
incoming_grouped=trips_from_abroad.groupby("origin")["trips"].sum().reset_index()

In [38]:
outgoing_grouped=outgoing_grouped.sort_values(by="trips",ascending=False)

In [36]:
incoming_grouped=incoming_grouped.sort_values(by="trips",ascending=False)

In [39]:
outgoing_grouped

destination      trips
36           FR  58908.511
85           PT  28366.809
37           GB  13100.902
105          US   4482.920
29           DE   4445.545
..          ...        ...
99           TD      3.599
98           SY      3.599
111          ZW      3.599
67           MK      1.587
58           LI      1.587

[112 rows x 2 columns]

In [40]:
incoming_grouped

origin      trips
37      FR  48896.216
84      PT  22186.364
38      GB  14478.282
50      IT   7282.493
29      DE   5934.856
..     ...        ...
53      KH      3.599
101     TT      3.599
97      SY      3.599
105     UZ      3.599
36      FO      2.732

[108 rows x 2 columns]